### PR TITLE
Speed up snapshotting of many new dom nodes

### PR DESCRIPTION
**Pull request by Juice10**
By avoiding reflow we shave about 15-25% off our snapshotting time.
By doing some other tweaks we can bring our savings up to about 40-50%.

Before:
image
image
image

After:
image
image
image

### CODE REVIEW

**Comment by Yuyz0112** [comment]
Very interesting! Going to learn the details.

**Comment by Juice10** [comment]
@Yuyz0112 I've finished tweaking this branch. Feel free to review it. Happy to answer any questions if you have them

**Comment by eoghanmurray** [comment]
:star_struck:

File: packages/rrweb-snapshot/src/snapshot.ts [hunk 1]
```diff
@@ -694,6 +482,290 @@ function serializeNode(
...
+  if (tagName === 'audio' || tagName === 'video') {
+    attributes.rr_mediaState = (n as HTMLMediaElement).paused
+      ? 'paused'
+      : 'played';
+    attributes.rr_mediaCurrentTime = (n as HTMLMediaElement).currentTime;
+  }
+  // Scroll
+  if (!newlyAddedElement) {
+    // `scrollTop` and `scrollLeft` are expensive calls because they trigger reflow.
+    // Since `scrollTop` & `scrollLeft` are always 0 when an element is added to the DOM.
```
Comment (Yuyz0112) [inline packages/rrweb-snapshot/src/snapshot.ts:729]: interesting approach 👍

**Review by Yuyz0112 — APPROVED** [review]

File: packages/rrweb/src/record/observers/canvas/webgl.ts [hunk 2]
```diff
@@ -31,8 +31,8 @@ function patchGLPrototype(
         return function (this: typeof prototype, ...args: Array<unknown>) {
           const result = original.apply(this, args);
           saveWebGLVar(result, win, prototype);
-          if (!isBlocked(this.canvas , blockClass)) {
-            const id = mirror.getId(this.canvas );
+          if (!isBlocked(this.canvas, blockClass, true)) {
+            // const id = mirror.getId(this.canvas);
```
[the remark this answers is not in the document]
Comment (Juice10) [reply packages/rrweb/src/record/observers/canvas/webgl.ts:35]: Absolutely
```suggestion
```

**Comment by Yuyz0112** [comment]
LGTM, we can merge this after conflicts are resolved.

**Comment by Juice10** [comment]
@Yuyz0112 fixed the merge conflicts. Feel free to merge this branch into master

File: packages/rrweb-snapshot/src/snapshot.ts (hunk 1 above)
Comment (eoghanmurray) [reply packages/rrweb-snapshot/src/snapshot.ts:729]: I'm wondering whether we could replace `newlyAddedElement` and `skipChild` with a single `fromMutation` argument as they are both true/false in the same places; I feel like that would suffice and be clearer, but let me know thoughts before I pursue that PR!

File: packages/rrweb-snapshot/test/snapshot.test.ts [hunk 3]
```diff
@@ -180,3 +180,41 @@ describe('style elements', () => {
...
+    });
+  };
+
+  const render = (html: string): HTMLDivElement => {
+    document.write(html);
+    return document.querySelector('div')!;
+  };
+
+  it('should serialize scroll positions', () => {
+    const el = render(`<div stylel='overflow: auto; width: 1px; height: 1px;'>
```
Comment (eoghanmurray) [inline packages/rrweb-snapshot/test/snapshot.test.ts:208]: was this a typo?  `stylel`

**Pull request merged**